training/src/main/scala/com/airbnb/aerosolve/training/photon/ml/data/ConvertExamplesToPhotonMLAvroJob.scala

```diff
@@ -0,0 +1,170 @@
+import com.airbnb.aerosolve.core.transforms.Transformer
+import com.airbnb.aerosolve.training.pipeline.GenericPipeline
+import com.typesafe.config.{Config, ConfigFactory, ConfigParseOptions, ConfigResolveOptions}
+import org.apache.avro.generic.{GenericData, GenericRecord}
+import org.apache.avro.mapred.AvroKey
+import org.apache.avro.mapreduce.{AvroJob, AvroKeyOutputFormat}
+import org.apache.avro.Schema
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+import org.apache.hadoop.io.NullWritable
+import org.apache.hadoop.mapreduce.Job
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.rdd.RDD
+import org.slf4j.LoggerFactory
+
+import scala.collection.JavaConverters._
+import scala.collection.immutable
+import scala.util.Try
+
+
+/**
+  * Providing a data conversion job casting Thrift [[com.airbnb.aerosolve.core.Example]] to PhotonML
+  * Avro format: https://github.com/linkedin/photon-ml#input-data-format
+  *
+  */
+object ConvertExamplesToPhotonMLAvroJob {
+
+  import PhotonMLUtils._
+
+  private val logger = LoggerFactory.getLogger(this.getClass())
+  private val LABEL = "label"
+
+  def main(args: Array[String]): Unit = {
+    val configPath = new File(args(0))
+
+    val config = if (configPath.exists()) {
+      ConfigFactory.load(ConfigFactory.parseFile(configPath))
+    } else {
+      ConfigFactory.load(configPath.getPath(),
+        ConfigParseOptions.defaults.setAllowMissing(false),
+        ConfigResolveOptions.defaults)
+    }
+
+    val sparkConf = new SparkConf().setAppName("Convert Thrift Example to PhotonML Avro")
+    val sc = new SparkContext(sparkConf)
+
+    convertExamplesToPhotonGAMEAvro(sc, config)
+  }
+
+  /**
+    * This method converts an RDD of examples to PhotonML GAME Avro formats and save to HDFS
+    *
+    * @param sc         The spark context
+    * @param taskConfig config of the task
+    */
+  def convertExamplesToPhotonGAMEAvro(sc: SparkContext,
+                                      taskConfig: Config): Unit = {
+    val (transformedData, avroSchema) = loadExamples(sc, taskConfig)
+
+    val outputPath = new Path(taskConfig.getString("output_path"))
+
+    val fs = outputPath.getFileSystem(new Configuration())
+    fs.delete(outputPath, true)
+
+    val avroSchemaStr = avroSchema.toString()
+    logger.info(s" Configured output Avro schema: ${avroSchema.toString(true)}")
+
+    val metaDataExtractorClass = Try(taskConfig.getString("meta_data_extractor_class"))
+      .getOrElse(classOf[SingleFloatFamilyMetaExtractor].getName())
+
+    val metaDataExtractorContext = Try(taskConfig
+      .getConfig("meta_data_extractor_context").entrySet().asScala
+      .map { case entry => (entry.getKey(), entry.getValue().toString()) }.toMap
+    ).getOrElse(immutable.Map[String, String]())
+
+    val trainingExampleAvros = transformedData.mapPartitions { case iter =>
+      val schema = new Schema.Parser().parse(avroSchemaStr)
```

Comment:
This is to prevent Avro schema object serialization issue. As far as I know, Avro internally are keeping some static state information, directly serializing the object is dangerous. So that's why I cast to string, and then deserialize it in remote executors.